### DISCUSSION TITLE
Inquiry: Support for External Inference Endpoint in Tabby

### DISCUSSION THREAD

**Discussion by d0rc** [TabbyML/tabby#767] [Ideas]
Hey Tabby Team,

Love what you're doing with Tabby, it's a game-changer! Quick question: is there a way to route Tabby's inference requests to an external endpoint? I've got a setup that batches requests to remote GPUs for better performance.

It could lower latency and let users like me use bigger, custom models. Plus, it might be a cool feature for others too. Just a thought – totally get it if it's not doable right now.

Thanks for all the awesome work!

**Comment by wsxiaoys** [comment]
Is this endpoint dedicate to tabby? If that's the case, you might consider deploy tabby to this machine directly?

**Reply by d0rc** [reply]
I was hoping I can give Tabby endpoint for inference, like with OpenAI protocol or something similar.

**Reply by wsxiaoys** [reply]
Related discussion: https://github.com/TabbyML/tabby/issues/795

**Comment by kinthaiofficial** [comment]
External inference endpoint support would make Tabby substantially more flexible in production environments where the model serving layer is already standardized. A few thoughts on the design:

**What the integration needs to handle:**
- OpenAI-compatible completions endpoint (`/v1/completions`) — most production inference stacks already implement this
- FIM (Fill-In-the-Middle) prompt template translation — Tabby's FIM requests need to map correctly to whatever format the external endpoint expects (starcoder format, codellama format, etc.)
- Streaming responses — IDE autocomplete needs low-latency first token, so streaming is non-negotiable

**Authentication to the external endpoint:**
API key passthrough in the config, loaded from env var. Don't hardcode credentials in the config file.

**Latency expectations:**
External inference adds network round-trip vs local inference. For autocomplete, this matters a lot — users feel latency >200ms as sluggish. Consider: request cancellation when the user types more characters before the response arrives.

**Fallback to local if external is unavailable:**
For resilience: try external endpoint first, fall back to local model if external returns non-200. This makes Tabby usable even when the inference endpoint is down for maintenance.

**Model capability negotiation:**
Different models support different context lengths and FIM formats. The config should let you specify which FIM template to use (or auto-detect if the endpoint exposes model metadata).

This feature would be especially valuable for teams that want unified model serving (one inference cluster, multiple tools consuming it) rather than per-tool model instances.
